File: Laboratory_4/GoLang/task5/strategies.go
package main

func enemyHasBetrayed(arr []bool) bool {
	for _, v := range arr {
		if !v {
			return true
		}
	}
	return false
}

func every23(round int, self []bool, enemy []bool) bool {
	if round%23 == 0 {
		return false
	}
	return true
}

func grimTrigger(round int, self []bool, enemy []bool) bool {
	if enemyHasBetrayed(enemy) {
		return false
	}
	return true
}

func titForTat(round int, self []bool, enemy []bool) bool {
	if len(enemy) == 0 {
		return true
	}
	return enemy[len(enemy)-1]
}

File: Laboratory_7/Laboratory_4/C++/task5/strategies.cpp
#include <vector>
#include "strategies.h"

using namespace std;

bool enemyHasBetrayed(vector<bool> arr) {
    /* Проверяет на предательство */
    for (int i = 0; i < arr.size(); i++) {
        if (!arr[i]) {
            return true;
        }
    }
    return false;
}

bool every23(int round_number, vector<bool> self_choices, vector<bool> enemy_choices) {
    /* Прредавать каждый 23 раунд */
    if (round_number % 23 == 0) {
        return false;
    }
    return true;
}

bool grimTrigger(int round_number, vector<bool> self_choices, vector<bool> enemy_choices) {
    /* Никогда не сотрудничать с тем, кто предал */
    if (enemyHasBetrayed(enemy_choices)) {
        return false;
    }
    return true;
}

bool titForTat(int round_number, vector<bool> self_choices, vector<bool> enemy_choices) {
    /* Сначала сотрудничать, потом делать так, как противник */
    if (enemy_choices.empty()) {
        return true;
    }
    return enemy_choices.back();
}